### ISSUE TITLE
Vectors layer not refreshing on refresh()

### ISSUE THREAD

**Issue by cnstt** [napari/napari#4694]
## 🐛 Bug

Vectors layer not refreshing despite having updated elements of the array.
EDIT: Vectors layer not refreshing on `refresh()`

## To Reproduce

```
import napari
import numpy as np
v = napari.Viewer()
data = np.array([[0,10],[0,5]])
viewer.add_vectors(data)
viewer.layers[0].data[0] = np.array([[0,10],[0,1]])
# Look at the data: it has been updated
viewer.layers[0].data
# But the viewer has not, event after triggering refresh
viewer.layers[0].refresh()
# But setting edge_width updates the viewer!
viewer.layers[0].edge_width=1
```

## Expected behavior

The viewer should update when changing parts of the array (because it does when you change the whole data at once) or at least the refresh() method should work...
EDIT: The viewer should update when triggering `refresh()`

## Environment
napari: 0.4.16
Platform: Linux-5.17.13-200.fc35.x86_64-x86_64-with-glibc2.34
System: Fedora Linux 35 (Workstation Edition)
Python: 3.9.13 (main, May 18 2022, 00:00:00) [GCC 11.3.1 20220421 (Red Hat 11.3.1-2)]
Qt: 5.15.2
PyQt5: 5.15.6
NumPy: 1.22.4
SciPy: 1.7.3
Dask: 2022.05.2
VisPy: 0.10.0

OpenGL:
- GL version: 4.6.0 NVIDIA 515.48.07
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1200, scale 1.0

Plugins:
- console: 0.0.4
- napari-deepfinder: 0.0.1
- napari-svg: 0.1.6
- scikit-image: 0.4.16

**Comment by Czaki** [comment]
The lack of refreshing on `refresh()` is a bug. But NumPy does not provide a way to know that values are updated.

**Comment by andy-sweet** [comment]
Hey @kcpevey - as you recently took a look over the vectors layer + vispy layer, thought this could be a good bug to try to fix.

**Comment by kcpevey** [comment]
Thanks for the bug report @cnstt ! 

---

TLDR;  
For the time being, the best workaround is to replace `data` completely (i.e. `viewer.layers[0].data = np.array([[0,10],[0,1]])`) rather than just changing a slice. 

---

I had a look and am able to reproduce the issue. 

As you noted, when you call `refresh()`, the visual display doesn't update, but when you call `viewer.layers[0].edge_width=1` (which then calls `refresh()`) _does_ change the visual display. 

The `edge_width` setter is calling `generate_vector_meshes` which updates `_mesh_vertices` and `_mesh_triangles` ([see this bit of code](https://github.com/napari/napari/blob/640484c10e2db29971130d296729445b52b3e98e/napari/layers/vectors/vectors.py#L452-L458)) which is what is being displayed. 

When `refresh()` is called directly, its using an old copy of `_mesh_vertices` and `_mesh_triangles` that is from the original data. `refresh()` calls `_set_view_slice` which does provide a way to call `generate_vector_meshes`, but only if the `dims_displayed` have changed ([see here](https://github.com/napari/napari/blob/640484c10e2db29971130d296729445b52b3e98e/napari/layers/vectors/vectors.py#L683)). If you have just changed a slice of the data and the dims haven't changed, then it doesn't regenerate the mesh. 

The issue here is that `refresh` is _also_ what's being called when we move the slider in the interface. If we were to add a `generate_vector_meshes` to every call of `_set_view_slice`, it would be a performance hit around O(100). 

@andy-sweet and I spent a few minutes thinking through options, but we didn't come up with any quick fixes. Some ideas were to pass a boolean to trigger the hard refresh, or splitting out the refresh call from the slider as a separate function. We didn't really like anything we came up with. 

@sofroniewn or @brisvag do you have any other ideas?

**Comment by brisvag** [comment]
Yup, that's exactly what I remembered about caching when this issue came up in the meeting. Thanks for digging out the details!

I agree with your assessment: there's no good way to solve this without changing strategy, because mesh computation is expensive and we want to minimize it. This is one of the reasons why I *really* wanted to get #4005 in before working on new slicing behaviour (#4005 was a dependency of the first iteration of thick slices in #4012).

My best answer is: we need to push forward #4005 and solve the macos and CI issues.

**Comment by sofroniewn** [comment]
Hmm, agreed - this isn't ideal. Some thoughts are - we might want to pull mesh generation out of the [napari/layers/vectors.py](https://github.com/napari/napari/blob/main/napari/layers/vectors/vectors.py) and put it in the [napari/_vispy/layers/vectors.py](https://github.com/napari/napari/blob/main/napari/_vispy/layers/vectors.py). What this does is then say that from napari's perspective how we actually render the vectors (do we use OpenGL lines or do we make our own meshes) becomes a rendering implementation detail and not something the layer itself cares about. (I'd probably propose that same rearchitecture for the shapes layer which also creates a lot of meshes). We could then think a bit better about caching those meshes and regenerating them only when needed, but it will be a simpler problem.

Curious what @brisvag @kcpevey @andy-sweet think about this idea? Feel free to say it's a really bad one!!!

**Comment by andy-sweet** [comment]
> Curious what @brisvag @kcpevey @andy-sweet think about this idea? Feel free to say it's a really bad one!!!

I like the idea in term of the semantics - the meshes should be more associated with the visualization/rendering (i.e. vispy) than the data/model (i.e. layer). But I don't immediately see how it would make solving this problem easier - we still have to work out when to regenerate them. Separating them might actually make things harder, though the vispy layer does have access to the napari layer, so maybe not.

I prefer the two other suggestions so far.

1. Get rid of the meshes entirely (i.e. #4005).
2. Only regenerate them in refresh, and don't use refresh for slicing (maybe done by the end of async slicing).

But could be convinced otherwise.

**Comment by brisvag** [comment]
I think I went down this rabbithole before (moving the mesh generation logic to the `VispyLayer` side), but found it even harder if not impossible, because you need to keep track of the viewd *slice*, which normally happens on the napari side... If someone can figure it out, I agree that it would be *theoretically* better, but I'd rather prioritize efforts towards #4005.

**Comment by sofroniewn** [comment]
So it might actually help a bit to move the mesh generation over to the vispy side as we'd only be generating meshes for the current slice. Right now we generate meshes for all the possible slices right when data is loaded - which is great once it's down as it means that slicing is very fast afterwards, but it could be terrible if you actually had lots and lots of slices (this is why we only regenerate the meshes when the dims displayed change but not the slices change).

I'm thinking @andy-sweet once we get to async slicing it will really be much nicer to have the mesh generation done per slice and on the vispy side.

> because you need to keep track of the viewd slice, which normally happens on the napari side... If someone can figure it out, I agree that it would be theoretically better,

I might take a look into this, see if i can do it - at least just for comparisons sake

**Comment by sofroniewn** [comment]
I gave this a quick attempt in #4846 - doesn't look too bad from a code standpoint, but I havn't done any profiling or anything else
